File: Pairly/android/app/src/main/java/com/pairly/PairlyWidgetProvider.java
package com.pairly;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.view.View;
import android.widget.RemoteViews;
import com.pairly.app.MainActivity; // Corrected import
import com.pairly.app.R; // Added import

import java.io.File;

public class PairlyWidgetProvider extends AppWidgetProvider {

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId, String photoPath, String partnerName, long timestamp) {
        try {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.pairly_widget_layout);

            if (photoPath != null && !photoPath.isEmpty()) {
                File imgFile = new File(photoPath);
                if (imgFile.exists()) {
                    Bitmap myBitmap = BitmapFactory.decodeFile(imgFile.getAbsolutePath());
                    views.setImageViewBitmap(R.id.widget_image, myBitmap);
                    views.setTextViewText(R.id.widget_partner_name, partnerName);
                    views.setViewVisibility(R.id.widget_placeholder, View.GONE);
                    views.setViewVisibility(R.id.widget_image, View.VISIBLE);
                    views.setViewVisibility(R.id.widget_partner_name, View.VISIBLE);
                } else {
                    // Image file does not exist, show placeholder
                    views.setViewVisibility(R.id.widget_placeholder, View.VISIBLE);
                    views.setViewVisibility(R.id.widget_image, View.GONE);
                    views.setViewVisibility(R.id.widget_partner_name, View.GONE);
                }
            } else {
                // No photo path, show placeholder
                views.setViewVisibility(R.id.widget_placeholder, View.VISIBLE);
                views.setViewVisibility(R.id.widget_image, View.GONE);
                views.setViewVisibility(R.id.widget_partner_name, View.GONE);
            }

            // Make the widget clickable to open the app
            Intent intent = new Intent(context, com.pairly.app.MainActivity.class); // Corrected class reference
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
            views.setOnClickPendingIntent(R.id.widget_root, pendingIntent);

            appWidgetManager.updateAppWidget(appWidgetId, views);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        // There may be multiple widgets active, so update all of them
        for (int appWidgetId : appWidgetIds) {
            // By default, show placeholder. The widget will be updated from the app.
            updateAppWidget(context, appWidgetManager, appWidgetId, null, null, 0);
        }
    }

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        // Enter relevant functionality for when the last widget is disabled
    }
}

File: Pairly/android/app/src/main/java/com/pairly/app/PairlyWidgetProvider.kt
package com.pairly.app

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.widget.RemoteViews
import java.io.File

class PairlyWidgetProvider : AppWidgetProvider() {

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        
        if (intent.action == AppWidgetManager.ACTION_APPWIDGET_UPDATE) {
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val widgetComponent = android.content.ComponentName(context, PairlyWidgetProvider::class.java)
            val widgetIds = appWidgetManager.getAppWidgetIds(widgetComponent)
            
            for (widgetId in widgetIds) {
                updateAppWidget(context, appWidgetManager, widgetId)
            }
        }
    }

    private fun updateAppWidget(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int
    ) {
        val views = RemoteViews(context.packageName, R.layout.widget_layout)
        
        // Get saved widget data
        val prefs = context.getSharedPreferences("pairly_widget", Context.MODE_PRIVATE)
        val photoPath = prefs.getString("photoPath", null)
        val partnerName = prefs.getString("partnerName", "Your Partner")
        
        // Set partner name
        views.setTextViewText(R.id.widget_partner_name, partnerName)
        
        // Load and set photo
        if (photoPath != null) {
            val photoFile = File(photoPath)
            if (photoFile.exists()) {
                try {
                    val bitmap = BitmapFactory.decodeFile(photoPath)
                    if (bitmap != null) {
                        views.setImageViewBitmap(R.id.widget_photo, bitmap)
                    } else {
                        views.setImageViewResource(R.id.widget_photo, R.drawable.widget_placeholder)
                    }
                } catch (e: Exception) {
                    views.setImageViewResource(R.id.widget_photo, R.drawable.widget_placeholder)
                }
            } else {
                views.setImageViewResource(R.id.widget_photo, R.drawable.widget_placeholder)
            }
        } else {
            views.setImageViewResource(R.id.widget_photo, R.drawable.widget_placeholder)
        }
        
        // Set click intent to open app
        val intent = Intent(context, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            context,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        views.setOnClickPendingIntent(R.id.widget_container, pendingIntent)
        
        appWidgetManager.updateAppWidget(appWidgetId, views)
    }
}
